### DISCUSSION TITLE
How to migrate to ONE_DIR?

### DISCUSSION THREAD

**Discussion by Nebucatnetzer** [docker-mailserver/docker-mailserver#3342] [Q&A] [answered]
I've been running DMS with the multi volume approach for quite a while now.
Are there any guides or hints on how to migrate to the ONE_DIR approach?
I haven't really found anything, just how to enable it but not how the directory structure looks like.

**Comment by polarathene** [comment, marked as answer]
We manage this all in one file, you can get a list of the state that is in separate locations at `/var/` that we relocate to `/var/mail-state/`:

https://github.com/docker-mailserver/docker-mailserver/blob/da8d3654b83eb78fdd2b22b1fe2c6b5eaf779869/target/scripts/startup/setup.d/mail_state.sh#L15-L36

If you're comfortable enough with reading the file, we have it commented to explain each step that is performed. Those locations substitute any extra `/` in the location with `-`, and symlink the original location to the new `/var/mail-state` location.

I haven't tested, so you might want to make a backup copy. If you create a volume bind mount for `/var/mail-state`, then the local folder for that should get all your existing mounted volumes related to `ONE_DIR` copied over. **Future runs will remove content from the original location**, assuming it's internal when `/var/mail-state` exists with the service directory being checked.

After that, we make sure the correct ownership and permissions is applied.

---

> Are there any guides or hints on how to migrate to the `ONE_DIR` approach?

I don't believe there is, as it became the default a while ago. I think some maintainers that wanted it didn't think there was many (if any) users still using separate mounts instead of consolidating into `/var/mail-state`.

Just look at the [current `compose.yaml`](https://github.com/docker-mailserver/docker-mailserver/blob/da8d3654b83eb78fdd2b22b1fe2c6b5eaf779869/compose.yaml) we have as an example config. You should now have:
- Mail data (`/var/mail`), your actual mailbox data which Dovecot manages.
- Mail state (`/var/mail-state`), internal state from any services/features enabled (_eg: runtime created data that services need to persist to function correctly_)
- Mail logs (`/var/log/mail`), optional location for persisting logs.
- DMS config (`/tmp/docker-mailserver`), any special config files we support belong here
- `/etc/localtime:ro`, optional if you'd like the container timezone to aligned to the host systems.

If you have `domainname` set, you probably want to merge that into your `hostname` field now. It shouldn't change anything, but we've been advising that for the past 2 years I think (_`domainname` is actually for a NIS domain, not what most thought it was for_).
